server/game/abilities/keyword/AmbushAbility.ts

```diff
@@ -1,36 +1,54 @@
 import AbilityHelper from '../../AbilityHelper';
+import { AbilityContext } from '../../core/ability/AbilityContext';
 import TriggeredAbility from '../../core/ability/TriggeredAbility';
+import { TriggeredAbilityContext } from '../../core/ability/TriggeredAbilityContext';
 import { Card } from '../../core/card/Card';
+import { UnitCard } from '../../core/card/CardTypes';
 import { KeywordName } from '../../core/Constants';
 import Game from '../../core/Game';
 import * as Contract from '../../core/utils/Contract';
 import { ITriggeredAbilityProps } from '../../Interfaces';
 
-/** @deprecated this is still WIP until attack changes are done */
-// export class AmbushAbility extends TriggeredAbility {
-//     public override readonly keyword: KeywordName | null = KeywordName.Ambush;
+export class AmbushAbility extends TriggeredAbility {
+    public override readonly keyword: KeywordName | null = KeywordName.Ambush;
 
-//     public static buildAmbushAbilityProperties(): ITriggeredAbilityProps {
-//         return {
-//             title: 'Ambush',
-//             optional: true,
-//             when: { onUnitEntersPlay: (event, context) => event.card === context.source },
-//             immediateEffect: [
-//                 AbilityHelper.immediateEffects.ready(),
-//                 AbilityHelper.immediateEffects.attack((context) => {
-//                     return { attacker: context.source };
-//                 })
-//             ]
-//         };
-//     }
+    public static buildAmbushAbilityProperties<TSource extends Card = Card>(): ITriggeredAbilityProps<TSource> {
+        return {
+            title: 'Ambush',
+            optional: true,
+            when: { onUnitEntersPlay: (event, context) => event.card === context.source },
+            immediateEffect: AbilityHelper.immediateEffects.conditional({
+                condition: AmbushAbility.unitWouldHaveAmbushTarget<TSource>,
+                onTrue: AbilityHelper.immediateEffects.sequential([
+                    AbilityHelper.immediateEffects.ready(),
+                    AbilityHelper.immediateEffects.attack((context) => {
+                        return {
+                            attacker: context.source,
+                            targetCondition: (card) => !card.isBase(),
+                            optional: false     // override the default optional behavior - once we've triggered ambush, the attack is no longer optional
+                        };
+                    })]),
+                onFalse: AbilityHelper.immediateEffects.noAction()
+            })
+        };
+    }
 
-//     public constructor(game: Game, card: Card) {
-//         if (!Contract.assertTrue(card.isUnit())) {
-//             return;
-//         }
+    private static unitWouldHaveAmbushTarget<TSource extends Card = Card>(context: TriggeredAbilityContext<TSource>): boolean {
```

Comment:
Is this the best way to do this? I haven't seen us set up an ability like this before

Reply:
To do what specifically? Can you elaborate on what the concern is about

Reply:
Two things:

1. It's not just checking for the presence of an enemy unit, per the specific rule it's checking if there is a unit that this unit can attack. So that means checking for cases like e.g. if there's a Sabine unit alone in the ground arena but her ability is live b/c of space arena units, the opponent can't do a ground ambush b/c there's no attackable ground unit.

Those rules get pretty involved so if we tried to rewrite them here we'd just be rewriting or separating out all of the checks in the attack system into a helper. The latter would be the more okay of the two but even then, the helper would need to have passed in all of the same data that goes into making an AttackSystem so it kind of gets to the point of asking what would be better about it. The biggest advantage would be we're not creating a transient object, but idk if that's worth the trade.

2. We do quietly use this pattern in a places already. If you look at the PlayUpgrade action, you can see it's including an `attachUpgrade` system call that it passes to super() as a target resolver so the ability knows which targets are legal to play the upgrade on: https://github.com/AMMayberry1/forceteki/blob/f34c6f076e4d3bb09285aacab91b88ff9cfaa2ea/server/game/actions/PlayUpgradeAction.ts#L15

So overall there's probably room to optimize a little to make this cleaner but I think doing that in a way that doesn't just shove the complexity somewhere else would take a lot of work, which I have trouble justifying given how infrequently we need to do this.